proposals/p2029.md

```diff
@@ -0,0 +1,103 @@
+must ensure it respects these reserved identifiers to enable seamless and
+error-free interoperability with C++. Equally, it would be prudent to define a
+set of identifiers which are reserved for internal Carbon implementation usage.
+
+## Background
+
+The full set of C++ reserved identifiers is listed
+[here](https://en.cppreference.com/w/cpp/language/identifiers). This proposal is
+solely considering the set of user-definable identifiers which are considered
+names in C++:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved;
+-   Identifiers that begin with an underscore are reserved in the global
+    namespace.
+
+Considerations for reserved keywords are discussed in
+[p0093](https://github.com/carbon-language/carbon-lang/pull/93).
+
+## Proposal
+
+Carbon reserves the following identifiers in the specified contexts:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved.
+
+Carbon toolchains are expected to emit an error when a reserved identifier is
+encountered. This will remove the risk of undefined / unexpected runtime
+behaviour, or name resolution clashes when linking with C++ libraries.
+
+## Details
+
+This proposal makes use of three assumptions:
+
+1. A Carbon package exporting a C++ header will be scoped in a C++ namespace
+   following the package name;
+2. All C++ imports to Carbon will exist within the package `Cpp` which
+   represents the root of the C++ global namespace.
+3. The default Carbon package can not be exported to C++, following the
+   [no other library can import this library even from within the default package](/docs/design/README.md#files-libraries-packages)
+   rule.
+
+The stated reserved identifiers for Carbon ensure that there will be no
+user-defined identifier clashes between Carbon and C++ in either
+interoperability direction. As Carbon does not have a global namespace, and
+assumption 2 above explicitly gives the C++ global namespace a named scope
+within Carbon, there is no need to consider a Carbon equivalent of the C++
+_Identifiers that begin with an underscore are reserved in the global namespace_
+reservation.
+
+## Rationale
+
+This proposal supports the
+[Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+goal. It also provides a reserved identifier space for Carbon internal
+implementation specifics.
+
+## Alternatives considered
+
+It was considered to not place any user-visible restrictions on identifiers, and
+handle conflict resolution or avoidance at interop boundaries. This was
+dismissed as this would require automatic re-writing of reserved-in-C++
+identifiers, deferring identification of name resolution issues to link time, or
+allowing undefined behaviour to occur.
```

Comment:
Unless Carbon reserves a superset of all C++ reserved identifiers and keywords, we still have to deal with this issue. Moreover, the C++ keyword set will increase over time, so we will have to do something when that happens. For example, if a Carbon program uses the identifier `assume` and C++23 makes that a keyword, then we will have an interop conflict.

I am not convinced that we should be taking the approach of reserving all identifiers in Carbon that might be problematic for C++ interoperability. I think it's reasonable to instead tell Carbon developers to not use identifiers that can't be named from C++ in code that wants to interoperate with C++. There would be benefit in not reserving those identifiers, because the C++ restrictions do cause problems in practice: in particular, the restriction for identifiers beginning with underscore-capital and the restriction for identifiers containing a `__` not at the beginning affect code that people want to write reasonably often.

I think we should seriously consider the option of not having any reserved identifiers in Carbon, and see where that causes problems and how we might address those problems. For example:
- Can implementations provide non-standard keywords? How? (Maybe we should have a `k#foo` syntax for opting into keyword behavior, or maybe we should reserve a leading `__`. Explorer currently has some `__keyword`s for some experimental features that haven't been through the design proposal process.)
- How is built-in functionality exposed? (See #1804 for my thoughts on that.)
- Do we provide a mechanism to change the name of a Carbon entity when exposed to C++, and if so, how? (My preference here is to say no, and that if you want this behavior you write an alias on the Carbon side that will then be available in both languages -- or enable `-fms-extensions` in Clang and use `__identifier(whatever)` as the C++ analogue of Carbon's `r#whatever`.)